documentation/testing_plans.md

```diff
@@ -0,0 +1,825 @@
+
+  return({
+    'agents'    => $agents,
+    'agentless' => $agentless
+  })
+}
+```
+
+As a plan author, you might want to ensure the `patch::get_targets` sub-plan
+correctly returns a hash with targets grouped by `puppet-agent` status, and then
+run the appropriate task on each group of targets. To do so, you can write a few
+tests that assert the following behavior:
+
+- The sub-plan groups targets by their `puppet-agent` status and returns a hash
+  with these groups.
+
+- The sub-plan returns an empty array for a group that has no member targets.
+
+- The plan runs the `patch` task on targets with agents.
+
+- The plan does not run the `patch` task if there are no targets with agents.
+
+- The plan runs the `puppet_agent::install` task on agentless targets.
+
+- The plan does not run the `puppet_agent::install` task if there are no
+  agentless targets.
+
+The plan and sub-plan both run tasks in the `puppet_agent` module. To ensure
+this module is available to the tests, you can add it to the `.fixtures.yml`
+file in the module's root directory. Because the `puppet_agent` module also uses
+files from the `facts` module, include that module as well.
+
+```yaml
+# .fixtures.yml
+fixtures:
+  forge_modules:
+    puppet_agent:
+      repo: "puppetlabs/puppet_agent"
+      ref: "4.4.0"
+    facts:
+      repo: "puppetlabs/facts"
+      ref: "1.4.0"
+```
+
+Because the plan invokes the `get_targets` function, which accesses Bolt's
+inventory to retrieve `Target` objects, the tests must include an inventory. You
+can override the default inventory in `BoltSpec` using the `inventory_data`
+function.
+
+By default, `BoltSpec` tests run in `execute_any_plan` mode, which means the
+`patch::get_targets` plan is run in each test. For tests where it is not
+necessary to make assertions about the sub-plan, you can stub the sub-plan and
+have it return a set value.
+
+Save the following tests to `patch/spec/plans/init_spec.rb`:
+
+```ruby
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'bolt_spec/plans'
+
+describe 'patch' do
+  include BoltSpec::Plans
+
+  # Inventory that is loaded by Bolt with targets that can be referenced
+  # by name in stubs and mocks
+  def inventory_data
+    {
+      'groups' => [
+        {
+          'name'    => 'agents',
+          'targets' => ['agent-target']
+        },
+        {
+          'name'    => 'agentless',
+          'targets' => ['agentless-target']
+        },
+        {
+          'name'    => 'empty',
+          'targets' => []
+        }
+      ]
+    }
+  end
+
+  # Configure Puppet and Bolt before running any tests
+  before(:all) do
+    BoltSpec::Plans.init
+  end
+
+  # Set the return value of puppet_agent::version for each target
+  # Stub the patch and puppet_agent::install tasks
+  before(:each) do
+    allow_task('puppet_agent::version').return_for_targets({
+      'agent-target'     => { 'version' => '7.0.0' },
```

Comment:
This can only stub `stdout` or `stderr` :thinking: it's a bit of a pickle since we still want the task to be runnable on its own too...you could maybe try `puppet --version` and filter on if that fails or not?

Reply:
Hm, they're failing for me locally and if I'm reading this right it applies to `return_for_targets` as well https://github.com/puppetlabs/bolt/blob/main/lib/bolt_spec/plans.rb#L106

Reply:
That says for command and script, while this is for the `allow_task` stub. Further down in that comment we also have several examples of task stubs/mocks returning values without `stdout` or `stderr` keys.

I'm not sure why these are failing for you since I'm able to run them successfully locally...